File: backend/db/repositories/user-repository.ts
/**
 * User Repository
 * Data access layer for user operations
 */

import { sql } from '../connection.ts';
import type { UserData, CreateUserRequest, UpdateUserRequest } from "@shared/types/api.types";

export class UserRepository {
  /**
   * Find user by ID
   */
  async findById(id: string): Promise<UserData | null> {
    try {
      const result = await sql`
        SELECT id, email, name, role, job_role, created_at
        FROM users
        WHERE id = ${id}
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding user by ID:', error);
      throw error;
    }
  }

  /**
   * Find user by email
   */
  async findByEmail(email: string): Promise<(UserData & { password: string }) | null> {
    try {
      const result = await sql`
        SELECT id, email, name, role, job_role, password, created_at
        FROM users
        WHERE email = ${email}
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding user by email:', error);
      throw error;
    }
  }

  /**
   * Get all users
   */
  async findAll(): Promise<UserData[]> {
    try {
      const result = await sql`
        SELECT id, email, name, role, job_role, created_at
        FROM users
        ORDER BY role, name
      `;
      return result;
    } catch (error) {
      console.error('Error fetching all users:', error);
      throw error;
    }
  }

  /**
   * Create new user
   */
  async create(userData: CreateUserRequest): Promise<UserData> {
    try {
      const result = await sql`
        INSERT INTO users (email, name, password, role, job_role)
        VALUES (${userData.email}, ${userData.name}, ${userData.password}, ${userData.role}, ${userData.job_role || null})
        RETURNING id, email, name, role, job_role, created_at
      `;
      return result[0];
    } catch (error) {
      console.error('Error creating user:', error);
      throw error;
    }
  }

  /**
   * Update user
   */
  async update(id: string, userData: UpdateUserRequest): Promise<UserData> {
    try {
      const updates: string[] = [];
      const values: any[] = [];

      if (userData.name !== undefined) {
        updates.push(`name = $${values.length + 1}`);
        values.push(userData.name);
      }
      if (userData.email !== undefined) {
        updates.push(`email = $${values.length + 1}`);
        values.push(userData.email);
      }
      if (userData.role !== undefined) {
        updates.push(`role = $${values.length + 1}`);
        values.push(userData.role);
      }
      if (userData.job_role !== undefined) {
        updates.push(`job_role = $${values.length + 1}`);
        values.push(userData.job_role);
      }

      if (updates.length === 0) {
        throw new Error('No fields to update');
      }

      updates.push('updated_at = NOW()');
      values.push(id);

      const result = await sql`
        UPDATE users
        SET ${sql(updates.join(', '))}
        WHERE id = ${id}
        RETURNING id, email, name, role, job_role, created_at, updated_at
      `;

      return result[0];
    } catch (error) {
      console.error('Error updating user:', error);
      throw error;
    }
  }

  /**
   * Update user password
   */
  async updatePassword(id: string, hashedPassword: string): Promise<void> {
    try {
      await sql`
        UPDATE users
        SET password = ${hashedPassword}, updated_at = NOW()
        WHERE id = ${id}
      `;
    } catch (error) {
      console.error('Error updating password:', error);
      throw error;
    }
  }

  /**
   * Delete user
   */
  async delete(id: string): Promise<void> {
    try {
      await sql`
        DELETE FROM users
        WHERE id = ${id}
      `;
    } catch (error) {
      console.error('Error deleting user:', error);
      throw error;
    }
  }

  /**
   * Bulk create users
   */
  async bulkCreate(users: CreateUserRequest[]): Promise<UserData[]> {
    try {
      const values = users.map(u =>
        `('${u.email}', '${u.name}', '${u.password}', '${u.role}', ${u.job_role ? `'${u.job_role}'` : 'NULL'})`
      ).join(', ');

      const result = await sql`
        INSERT INTO users (email, name, password, role, job_role)
        VALUES ${sql.raw(values)}
        RETURNING id, email, name, role, job_role, created_at, updated_at
      `;

      return result;
    } catch (error) {
      console.error('Error bulk creating users:', error);
      throw error;
    }
  }

  /**
   * Check if email exists
   */
  async emailExists(email: string): Promise<boolean> {
    try {
      const result = await sql`
        SELECT COUNT(*) as count
        FROM users
        WHERE email = ${email}
      `;
      return result[0].count > 0;
    } catch (error) {
      console.error('Error checking email existence:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const userRepository = new UserRepository();

File: backend-python/app/repositories/user_repository.py
"""
User Repository
Database operations for user entities
"""

import logging
from typing import Any
from uuid import uuid4

from app.database import fetch, fetchone, execute, record_to_dict, records_to_list
from app.models.user import UserData, UserWithPassword, CreateUserRequest, UpdateUserRequest

logger = logging.getLogger(__name__)


class UserRepository:
    """Repository for user database operations."""

    async def find_by_id(self, user_id: str) -> UserData | None:
        """Find user by ID."""
        try:
            query = """
                SELECT id, name, email, role, job_role, created_at
                FROM users WHERE id = $1
            """
            record = await fetchone(query, user_id)
            if not record:
                return None
            return UserData(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error finding user by ID: {e}")
            raise

    async def find_by_email(self, email: str) -> UserWithPassword | None:
        """Find user by email (includes password for auth)."""
        try:
            query = """
                SELECT id, name, email, password, role, job_role, created_at
                FROM users WHERE email = $1
            """
            record = await fetchone(query, email.lower())
            if not record:
                return None
            return UserWithPassword(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error finding user by email: {e}")
            raise

    async def find_all(self) -> list[UserData]:
        """Get all users."""
        try:
            query = """
                SELECT id, name, email, role, job_role, created_at
                FROM users ORDER BY created_at DESC
            """
            records = await fetch(query)
            return [UserData(**r) for r in records_to_list(records)]
        except Exception as e:
            logger.error(f"Error fetching all users: {e}")
            raise

    async def create(self, user_data: CreateUserRequest | dict[str, Any]) -> UserData:
        """Create a new user."""
        try:
            if isinstance(user_data, CreateUserRequest):
                data = user_data.model_dump()
            else:
                data = user_data

            user_id = str(uuid4())
            query = """
                INSERT INTO users (id, name, email, password, role, job_role)
                VALUES ($1, $2, $3, $4, $5, $6)
                RETURNING id, name, email, role, job_role, created_at
            """
            record = await fetchone(
                query,
                user_id,
                data["name"],
                data["email"].lower(),
                data["password"],
                data.get("role", "learner"),
                data.get("job_role"),
            )
            return UserData(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error creating user: {e}")
            raise

    async def update(self, user_id: str, user_data: UpdateUserRequest | dict[str, Any]) -> UserData:
        """Update a user."""
        try:
            if isinstance(user_data, UpdateUserRequest):
                data = user_data.model_dump(exclude_unset=True)
            else:
                data = {k: v for k, v in user_data.items() if v is not None}

            # Build dynamic update query
            update_parts = []
            params = [user_id]
            param_idx = 2

            for key, value in data.items():
                if key == "email" and value:
                    value = value.lower()
                update_parts.append(f"{key} = ${param_idx}")
                params.append(value)
                param_idx += 1

            if not update_parts:
                # No updates, just return current user
                return await self.find_by_id(user_id)

            query = f"""
                UPDATE users SET {', '.join(update_parts)}
                WHERE id = $1
                RETURNING id, name, email, role, job_role, created_at
            """
            record = await fetchone(query, *params)
            if not record:
                raise ValueError("User not found")
            return UserData(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error updating user: {e}")
            raise

    async def delete(self, user_id: str) -> None:
        """Delete a user."""
        try:
            query = "DELETE FROM users WHERE id = $1"
            await execute(query, user_id)
        except Exception as e:
            logger.error(f"Error deleting user: {e}")
            raise

    async def update_password(self, user_id: str, hashed_password: str) -> None:
        """Update user password."""
        try:
            query = "UPDATE users SET password = $2 WHERE id = $1"
            await execute(query, user_id, hashed_password)
        except Exception as e:
            logger.error(f"Error updating password: {e}")
            raise

    async def email_exists(self, email: str) -> bool:
        """Check if email exists."""
        try:
            query = "SELECT EXISTS(SELECT 1 FROM users WHERE email = $1)"
            result = await fetchone(query, email.lower())
            return result[0] if result else False
        except Exception as e:
            logger.error(f"Error checking email existence: {e}")
            raise

    async def bulk_create(self, users: list[CreateUserRequest | dict[str, Any]]) -> list[UserData]:
        """Bulk create users."""
        created_users = []
        for user_data in users:
            try:
                user = await self.create(user_data)
                created_users.append(user)
            except Exception as e:
                logger.error(f"Error creating user in bulk: {e}")
                # Continue with other users
        return created_users


# Singleton instance
user_repository = UserRepository()
